### PR TITLE
add 4d normalize support

### PR DESCRIPTION
**Pull request by SeaOfOcean**
## What changes were proposed in this pull request?

add 4d normalize support, add optional scale operation

## How was this patch tested?

unit test

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala [hunk 1]
```diff
@@ -43,9 +47,16 @@ class Normalize[T: ClassTag](val p: Double, val eps: Double = 1e-10)
...
   var indices = Tensor[T]()
 
+  var cmul: CMul[T] = null
+
   override def updateOutput(input: Tensor[T]): Tensor[T] = {
-    require(input.dim() <= 2, s"Normalize: only 1d layer supported, " +
-      s"but got input dim ${input.dim()}")
+    require(input.dim() <= 2 || input.dim() == 4, s"Normalize: only 1d or 4d layer supported, " +
+      s"but got input dim ${ input.dim() }")
+    if (scale.isDefined) require(input.dim() == 4, s"currently only 4d layer supported scale")
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala]: Why `scale` is not supported for 1D input?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala [hunk 2]
```diff
@@ -62,24 +73,32 @@ class Normalize[T: ClassTag](val p: Double, val eps: Double = 1e-10)
...
       norm.resizeAs(normp).pow(normp, ev.fromType(1.0 / p))
     }
-    output.cdiv(inputBuffer, norm.view(norm.nElement(), 1).expandAs(inputBuffer))
+    if (norm.dim() <= 2) {
+      output.cdiv(inputBuffer, norm.view(norm.nElement(), 1).expandAs(inputBuffer))
+    } else if (norm.dim() == 4) {
+      output.cdiv(inputBuffer, norm.view(norm.size()).expandAs(inputBuffer))
+    }
+    if (scale.isDefined) {
+      output = cmul.forward(output)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala]: Is cmul trainable?

Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala]: yes, we want scales to be trainable

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala [hunk 3]
```diff
@@ -18,17 +18,21 @@ package com.intel.analytics.bigdl.nn
 import com.intel.analytics.bigdl.nn.abstractnn.TensorModule
 import com.intel.analytics.bigdl.tensor.Tensor
 import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{T, Table}
 
 import scala.reflect.ClassTag
 
 /**
  * Normalizes the input Tensor to have unit L_p norm. The smoothing parameter eps prevents
  * division by zero when the input contains all zero elements (default = 1e-10).
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala]: done

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala [hunk 4]
```diff
@@ -45,12 +48,14 @@ class Normalize[T: ClassTag](val p: Double, val eps: Double = 1e-10,
   var inputBuffer = Tensor[T]()
 
   var cross = Tensor[T]()
+  var crossBuffer = Tensor[T]()
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala]: clear this buffer in clearState

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala [hunk 5]
```diff
@@ -143,7 +150,10 @@ class Normalize[T: ClassTag](val p: Double, val eps: Double = 1e-10,
     gradInput.cdiv(cross.expandAs(inputBuffer))
     gradInput = gradInput.view(input.size())
     if (scale.isDefined) {
-      gradInput = cmul.backward(output, gradInput)
+      cmul.gradInput = cmul.backward(output, gradInput)
+      val tmp = cmul.gradInput
+      cmul.gradInput = gradInput
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala]: why?

**Comment by SeaOfOcean** [comment]
pr validate pass

**Comment by qiuxin2012** [comment]
Please merge into one commit.

**Pull request merged**